DM-40451: Allow deblending to continue even when data is missing for some bands

fred3m opened the pull request:
Previously any footprints that could not construct a PSF in all bands were skipped. As part of HSC PDR4, because there are patches with only partial coverage of narrow bands (for example), we want to allow the deblender to proceed.

This ticket implements deblending with only partial PSF coverage, and also adds new columns to the catalog to keep track of the coverage per band.

fred3m committed:
Allow deblending to continue even when data is missing for some bands

Previously any footprints that could not construct a PSF in all
bands were skipped. As part of HSC PDR4, because there are patches
with only partial coverage of narrow bands (for example), we want
to allow the deblender to proceed.

This ticket implements deblending with only partial PSF coverage,
and also adds new columns to the catalog to keep track of the
coverage per band.

cmsaunders commented on python/lsst/meas/extensions/scarlet/io.py:
```diff
@@ -422,7 +444,35 @@ def updateCatalogFootprints(self, catalog, band, psfModel, redistributeImage=Non
+        The fraction of pixels in `footprint` where the ``NO_DATA`` bit is set.
+    """
+    # Store the value of "NO_DATA" from the mask plane.
+    noDataInt = 2**maskImage.getMaskPlaneDict()["NO_DATA"]
```
You could just do `noDataInt = maskImage.getPlaneBitMask("NO_DATA")` here.

fred3m replied:
Oooo, nice!

cmsaunders commented on python/lsst/meas/extensions/scarlet/scarletDeblendTask.py:
```diff
@@ -238,7 +268,7 @@ def deblend(mExposure, footprint, config, spectrumInit):
     blend.psfCenter = (psfCenter.x, psfCenter.y)
 
-    return blend, skipped
+    return blend, skipped, []
```
Can you add a comment explaining this? Also, I saw that the docstrings say `spectrumInit` is returned, but it looks like it's not. 
Edit: Okay, I see that you have this to match the output of `deblend_lite`, but a comment plus corrected docstrings would still be helpful.

cmsaunders commented on python/lsst/meas/extensions/scarlet/scarletDeblendTask.py:
```diff
@@ -108,6 +108,36 @@ def isPseudoSource(source, pseudoColumns):
+
+    try:
+        psfModels = mExposure.computePsfKernelImage(psfCenter)
+    except IncompleteDataError as e:
```
My general feeling is that try-excepts should be avoided as much as possible. Since you are already changing the parameters and returned values in `computePsfKernelImage`, you could consider adding something like an `allowMissing` option, then always returning the psfs and the filters used. I'm not sure if that would be simpler.

fred3m replied:
I think this is a good example of a case where it is easier to ask forgiveness than acceptance. I originally used an `allowMissing` parameter but it requires an additional check by the user to see if any bands were missing. Since this check is already being done while creating the PSF image I wanted to find a way to avoid having to validate the output again, and this seemed like a good way to ensure that users who don't expect a bad PSF will get an error, and users who know that it might happen do not have to check for it.

cmsaunders replied:
Ok, sounds reasonable to me.

cmsaunders commented on python/lsst/meas/extensions/scarlet/scarletDeblendTask.py:
```diff
@@ -819,6 +865,10 @@ def _addSchemaKeys(self, schema):
         self.deblendErrorKey = schema.addField('deblend_error', type="String", size=25,
                                                doc='Name of error if the blend failed')
+        self.incompleteDataKey = schema.addField('deblend_incompleteData', type='Flag',
+                                                 doc='True when a bland has at least one band '
```
`bland` -> `blend`

cmsaunders commented on tests/test_deblend.py:
```diff
@@ -104,6 +104,7 @@ def test_deblend_task(self):
         src.setFootprint(denseFoot)
 
         # Run the deblender
+        print("RUNNING!")
```
I assume this was for debugging?

cmsaunders commented on python/lsst/meas/extensions/scarlet/scarletDeblendTask.py:
```diff
@@ -1009,9 +1061,9 @@ def deblend(self, mExposure, catalog):
                 # Build the parameter lists with the same ordering
                 if self.config.version == "scarlet":
-                    blend, skipped = deblend(mExposure, foot, self.config, spectrumInit)
+                    blend, skippedSources, skippedBands = deblend(mExposure, foot, self.config, spectrumInit)
```
Continuing the thought from above, you could have `deblend` just return `blend` and `skippedSources`, then set `skippedBands = []` here, so that the `deblend` outputs aren't forced to match `deblend_lite`.

fred3m replied:
Yeah, that does make more sense.

cmsaunders reviewed (approved):
This looks good. I have a few comments about places where I think you could make the logic more straightforward, but it's your call whether you make those changes.

fred3m committed:
Respond to reviewer comments

fred3m committed:
Make imageMask optional for reconstructing footprints from scarlet models

Because deepCoadd_forced does not require the imageMask to set the
coverage column (which instead is in the meas catalog),
this ticket makes that parameter optional.
